[build] Update compilation flags and fix all warnings

cbentejac opened the pull request:
This PR adds the following recommended compilation flags to all the QtAliceVision plugins (QtAliceVision, QtAliceVisionImageIO and DepthMapEntity):  `-Wall -Wextra -Wconversion -Wsign-conversion  -Wshadow -Wpedantic`.

All the warnings raised by these new flags (as well as those that existed prior to the addition of these flags) are fixed. Additionally, warnings raised by the `-Wswitch-default` and `-Wold-style-cast` flags are also fixed, although the flags themselves are not added to the plugins' compilation.

Most of the warnings that have been fixed consisted in:
- implicit casts (and C-style casts)
- incorrect syntax for Q_PROPERTY declarations
- calls to deprecated functions in OpenImageIO and Qt3D
- declaration of attributes in function prototypes that are needed but were never used

Finally, some minor clean-up has been performed in order to improve the readability: trailing spaces have been removed and lines of code that were very long, especially because of the explicit casting, have been split. Casts have also been limited when it was possible, which improves the code readability as well.

simogasp commented on src/MSfMData.hpp:
```diff
@@ -91,7 +91,7 @@ class MSfMData : public QObject
         if(!_sfmData || _status != Ready)
             return 0;
-        return _sfmData->getValidViews().size();
+        return (int)_sfmData->getValidViews().size();
```
Would it make more sense to return a std::size_t instead of int? I don't know if there are other constraints for using `int` but it seems more natural.

cbentejac replied:
Done! I think the `int` was there because it didn't require to be registered with Qt to be used in properties.

simogasp commented on src/MTracks.hpp:
```diff
@@ -96,7 +96,7 @@ class MTracks : public QObject
         if(!_tracks || _status != MTracks::Ready)
             return 0;
-        return _tracks->size();
+        return (int)_tracks->size();
```
same here, I'd rather return a std::size_t that represents a positive number.

cbentejac replied:
Done as well.

simogasp commented:
If it does not open a can of worms :-) it would be better to add a 
```cmake
if(MSVC)
    target_compile_options(qtAliceVisionPlugin PUBLIC /W4)
else()
    target_compile_options(qtAliceVisionPlugin PUBLIC -Wall -Wextra -Wconversion -Wsign-conversion  -Wshadow -Wpedantic ) # ideally, also -Werror
endif()
```
which are the recommended flags for c++ safe projects

cbentejac commented:
> If it does not open a can of worms :-) it would be better to add a
> 
> ```cmake
> if(MSVC)
>     target_compile_options(qtAliceVisionPlugin PUBLIC /W4)
> else()
>     target_compile_options(qtAliceVisionPlugin PUBLIC -Wall -Wextra -Wconversion -Wsign-conversion  -Wshadow -Wpedantic ) # ideally, also -Werror
> endif()
> ```
> 
> which are the recommended flags for c++ safe projects

So I compiled with these flags... and way too many warnings appeared. :)  I'll set this PR to draft to take the time to fix them all so we can use these flags in the long run.

cbentejac committed:
[build] Fix all MSVC default compilation warnings

- Implicit conversions from 'size_t' to 'int' or 'uint'
- Incompabilities between signed/unsigned variables with '<' operators
- Invalid Q_PROPERTY declaration (no READ accessor or associated MEMBER
variable)

cbentejac committed:
[build] Add extra-warning compilation flags for the QtAliceVision plugin

The -Werror flag is excluded from the list of added flags to still be
able to compile despite the warnings, should there ever be a need for
it.

cbentejac committed:
[build] Add extra-warning compilation flags for the DepthMapEntity plugin

cbentejac committed:
[build] Add extra-warning compilation flags for QtAliceVisionImageIO

cbentejac committed:
[build] Fix all warnings for the QtAliceVisionImageIO plugin

cbentejac committed:
[build] Fix all compilation warnings for the DepthMapEntity plugin

This fixes warnings for:
- the implicit casts, which are now explicit
- the use of deprecated functions in OpenImageIO and Qt3D
- the incorrect declaration of Q_PROPERTY
- the declaration of unused attributes that are needed

cbentejac committed:
[build] Fix all compilation warnings for the QtAliceVision plugin

Most of the fixes consist in:
- declaring necessary but unused attributes in function prototypes as such
- explicitly casting all variables that used to be implicitly cast
- using the correct syntax for Q_PROPERTY declarations

Some minor functions have been rewritten to fix the warnings.

cbentejac committed:
[build] DepthMapEntity: Fix all old-style casting warnings

These warnings are not raised by any of the compilation flags used in
the current build, but regularly appear in IDEs.

This commit fixes all the warnings that would have been raised with
the "-Wold-style-cast" compilation flag.

cbentejac committed:
[build] QtAliceVisionImageIO: Fix all old-style casting warnings

These warnings are not raised by any of the compilation flags used in
the current build, but regularly appear in IDEs.

This commit fixes all the warnings that would have been raised with
the "-Wold-style-cast" compilation flag.

cbentejac committed:
[build] QtAliceVision: Fix all old-style casting warnings

These warnings are not raised by any of the compilation flags used in
the current build, but regularly appear in IDEs.

This commit fixes all the warnings that would have been raised with
the "-Wold-style-cast" compilation flag.

cbentejac committed:
Clean-up: Remove trailing spaces

cbentejac committed:
[build] Fix all "switch missing default case" warnings

Add a "default" case to all the switches that were missing it.
This fixes all warnings raised by the "-Wswitch-default" compilation
flag.

cbentejac committed:
Reduce the number of casts when possible

For functions that cast the same variable to the same type repeatedly,
reduce the number of casts for that variable by casting it once into
a new variable, and using the new variable.

The goal is to both limit casts and improve the readability.

cbentejac committed:
Clean-up: Improve readability

Split long lines into several ones to improve the readability, especially
those that contain multiple casts.

simogasp commented on src/MFeatures.cpp:
```diff
@@ -697,21 +701,23 @@ void MFeatures::updatePerTrackInformation()
             ++trackFreatures.nbLandmarks;
-            if (trackIdPerLandmark.find(feature->landmarkId()) == trackIdPerLandmark.end())
-              trackIdPerLandmark[feature->landmarkId()] = feature->trackId();
+            unsigned int unsignedLandmarkId = static_cast<uint>(feature->landmarkId());
```
```suggestion
            const auto unsignedLandmarkId = static_cast<uint>(feature->landmarkId());
```

cbentejac replied:
Done.

simogasp commented on src/Surface.cpp:
```diff
@@ -166,20 +173,20 @@ void Surface::computeVerticesGrid(QSGGeometry::TexturedPoint2D* vertices, QSize
 
-            float u = i / (float)_subdivisions;
-            float v = j / (float)_subdivisions;
+            float u = fI / fSubdivisions;
+            float v = fJ / fSubdivisions;
```
not sure but they could be const
```suggestion
            const float u = fI / fSubdivisions;
            const float v = fJ / fSubdivisions;
```

cbentejac replied:
Indeed they can!

simogasp commented on src/qtAliceVisionImageIO/QtAliceVisionImageIOHandler.cpp:
```diff
@@ -141,6 +141,7 @@ bool QtAliceVisionImageIOHandler::read(QImage *image)
 bool QtAliceVisionImageIOHandler::write(const QImage &image)
 {
     // TODO
+    (void)image;  // Remove "unused parameter" warning
```
just remove the parameter name from the signature above:
```c++
bool QtAliceVisionImageIOHandler::write(const QImage&)
```

cbentejac replied:
Done.

simogasp reviewed (dismissed):
Sorry for this can of worms :-) But good job!

fabiencastan commented on src/depthMapEntity/plugin.hpp:
```diff
@@ -13,7 +13,12 @@ class DepthMapEntityPlugin : public QQmlExtensionPlugin
-    void initializeEngine(QQmlEngine* engine, const char* uri) override {}
+    void initializeEngine(QQmlEngine* engine, const char* uri) override
+    {
+        // Fix "unused parameter" warnings; should be replaced by [[maybe_unused]] when C++17 is supported
```
the alternative solution is to avoid putting a name in the arguments, like
like `void initializeEngine(QQmlEngine*, const char*) override {}`

fabiencastan commented on src/CMakeLists.txt:
```diff
@@ -49,6 +49,12 @@ find_package(Qt5 COMPONENTS Charts REQUIRED)
 # Target properties
 add_library(qtAliceVisionPlugin SHARED ${PLUGIN_SOURCES} ${PLUGIN_HEADERS} ${PLUGIN_MOCS})
 
+if(MSVC)
+    target_compile_options(qtAliceVisionPlugin PUBLIC /W4)
+else()
+    target_compile_options(qtAliceVisionPlugin PUBLIC -Wall -Wextra -Wconversion -Wsign-conversion -Wshadow -Wpedantic)
+endif()
+
```
Not sure, if it's a good idea to have that by default
